File: vm/src/main/java/com/libreshockwave/vm/trace/TracingHelper.java
package com.libreshockwave.vm.trace;

import com.libreshockwave.chunks.ScriptChunk;
import com.libreshockwave.vm.Datum;
import com.libreshockwave.vm.Scope;
import com.libreshockwave.vm.TraceListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Helper for building trace information and annotations.
 * Builds data structures for tracing - does not perform any I/O.
 * For console output, see {@link ConsoleTracePrinter}.
 */
public class TracingHelper {

    public TracingHelper() {
    }

    /**
     * Build instruction info for tracing.
     */
    public TraceListener.InstructionInfo buildInstructionInfo(Scope scope, ScriptChunk.Handler.Instruction instr) {
        String annotation = buildAnnotation(scope, instr);
        List<Datum> stackSnapshot = new ArrayList<>();
        // Capture up to 10 stack items
        for (int i = 0; i < Math.min(10, scope.stackSize()); i++) {
            stackSnapshot.add(scope.peek(i));
        }
        return new TraceListener.InstructionInfo(
            scope.getBytecodeIndex(),
            instr.offset(),
            instr.opcode().name(),
            instr.argument(),
            annotation,
            scope.stackSize(),
            stackSnapshot
        );
    }

    /**
     * Build handler info for tracing.
     */
    public TraceListener.HandlerInfo buildHandlerInfo(
            ScriptChunk script,
            ScriptChunk.Handler handler,
            List<Datum> args,
            Datum receiver,
            Map<String, Datum> globals) {
        return new TraceListener.HandlerInfo(
            script.getHandlerName(handler),
            script.id(),
            script.getDisplayName(),
            args,
            receiver,
            new HashMap<>(globals),
            script.literals(),
            handler.localCount(),
            handler.argCount()
        );
    }

    /**
     * Build annotation string for an instruction.
     * Uses the scope's script for name resolution (without local/param name resolution).
     */
    public String buildAnnotation(Scope scope, ScriptChunk.Handler.Instruction instr) {
        return InstructionAnnotator.annotate(scope.getScript(), instr);
    }
}

File: cpp/src/lingo/vm/trace/TracingHelper.cpp
#include "libreshockwave/lingo/vm/trace/TracingHelper.hpp"

#include <algorithm>
#include <string>
#include <utility>

#include "libreshockwave/lingo/Opcode.hpp"
#include "libreshockwave/lingo/vm/trace/InstructionAnnotator.hpp"

namespace libreshockwave::lingo::vm::trace {

TraceListener::InstructionInfo TracingHelper::buildInstructionInfo(
    const Scope& scope,
    const chunks::ScriptChunk::Instruction& instruction,
    const RuntimeGlobals& globals,
    const chunks::ScriptNamesChunk* names) const {
    std::vector<Datum> stackSnapshot;
    const int snapshotCount = std::min(10, scope.stackSize());
    stackSnapshot.reserve(static_cast<std::size_t>(snapshotCount));
    for (int index = 0; index < snapshotCount; ++index) {
        stackSnapshot.push_back(scope.peek(index).deepCopy());
    }

    return TraceListener::InstructionInfo{
        scope.bytecodeIndex(),
        instruction.offset,
        std::string(mnemonic(instruction.opcode)),
        instruction.argument,
        buildAnnotation(scope, instruction, names),
        scope.stackSize(),
        std::move(stackSnapshot),
        captureLocals(scope, names),
        globals,
    };
}

std::unordered_map<std::string, Datum> TracingHelper::captureLocals(
    const Scope& scope,
    const chunks::ScriptNamesChunk* names) const {
    std::unordered_map<std::string, Datum> locals;
    const auto* script = scope.script();
    if (script == nullptr) {
        return locals;
    }

    const auto& handler = scope.handler();
    for (int index = 0; index < static_cast<int>(handler.argNameIds.size()); ++index) {
        locals[script->resolveName(handler.argNameIds[static_cast<std::size_t>(index)], names)] =
            scope.getParam(index).deepCopy();
    }
    for (int index = 0; index < static_cast<int>(handler.localNameIds.size()); ++index) {
        locals[script->resolveName(handler.localNameIds[static_cast<std::size_t>(index)], names)] =
            scope.getLocal(index).deepCopy();
    }
    return locals;
}

TraceListener::HandlerInfo TracingHelper::buildHandlerInfo(
    const chunks::ScriptChunk& script,
    const chunks::ScriptChunk::Handler& handler,
    std::span<const Datum> args,
    const Datum& receiver,
    const RuntimeGlobals& globals,
    const chunks::ScriptNamesChunk* names,
    const std::string& scriptDisplayName) const {
    return TraceListener::HandlerInfo{
        script.getHandlerName(handler, names),
        script.id().value(),
        scriptDisplayName.empty() ? "script#" + std::to_string(script.id().value()) : scriptDisplayName,
        std::vector<Datum>(args.begin(), args.end()),
        receiver,
        globals,
        script.literals(),
        handler.localCount,
        handler.argCount,
    };
}

TraceListener::HandlerInfo TracingHelper::buildHandlerInfo(
    const chunks::ScriptChunk& script,
    const chunks::ScriptChunk::Handler& handler,
    const std::vector<Datum>& args,
    const Datum& receiver,
    const RuntimeGlobals& globals,
    const chunks::ScriptNamesChunk* names,
    const std::string& scriptDisplayName) const {
    return buildHandlerInfo(script, handler, std::span<const Datum>(args), receiver, globals, names, scriptDisplayName);
}

std::string TracingHelper::buildAnnotation(
    const Scope& scope,
    const chunks::ScriptChunk::Instruction& instruction,
    const chunks::ScriptNamesChunk* names) const {
    const auto* script = scope.script();
    if (script == nullptr) {
        return instruction.toString();
    }
    return InstructionAnnotator::annotate(*script, instruction, names);
}

} // namespace libreshockwave::lingo::vm::trace
